### PR TITLE
Improve error message when application description contains non-ASCII codepoints

### PR DESCRIPTION
**Pull request by antipax**
Improves the situation described in #4578.

@josevalim, you mentioned you wanted to investigate, so not sure if there were more similar issues you wanted to address, but this addresses the one mentioned in the issue, at least.

### CODE REVIEW

File: lib/mix/lib/mix/tasks/compile.app.ex [hunk 1]
```diff
@@ -160,6 +160,10 @@ defmodule Mix.Tasks.Compile.App do
         unless is_list(value) do
           Mix.raise "Application description (:description) is not a character list, got: #{inspect value}"
         end
+
+        unless Enum.all?(value, fn char -> char in 0..127 end) do
+          Mix.raise "Application description (:description) may contain only ASCII-printable codepoints, got: #{inspect value}"
```
Comment (lexmag) [inline lib/mix/lib/mix/tasks/compile.app.ex]: I think we should not say `printable` since we check for every ASCII code and they all are allowed.

Comment (antipax) [inline lib/mix/lib/mix/tasks/compile.app.ex]: Good call, I was reusing the wording about charlists from `List` but it's not really applicable in this case.

**Comment by lexmag** [comment]
I'm not sure about this change, seems we may have random Unicode codepoints in the `:description`:

```elixir
"うえ ειξήριολ øsdfh Й"
"うえ ειξήριολ øsdfh ЙSão"
"ЙSão"
```
Any of this works perfectly but not `São`. 😕

**Comment by antipax** [comment]
Yeah, I was a little weirded out myself. I tested with a description of `0..512 |> Enum.to_list` and it worked, but [227] does not. Seems safe enough to say only ASCII is allowed, but maybe it's worth investigating exactly what is causing the "bad encoding" error.

**Comment by antipax** [comment]
To add to the strangeness, I can't find "bad encoding" anywhere in the Erlang/OTP source code, except for [here](https://github.com/erlang/otp/blob/e1489c448b7486cdcfec6a89fea238d88e6ce2f3/lib/kernel/src/application_controller.erl#L1457) and [here](https://github.com/erlang/otp/blob/e1489c448b7486cdcfec6a89fea238d88e6ce2f3/lib/kernel/src/application_controller.erl#L1871), and those seem to be related to the name of the application rather than the description.

**Comment by antipax** [comment]
Also, the only specification for description [here](http://erlang.org/doc/man/app.html) is that it is of type `string()`, which is defined as a list of chars, chars being an integer from 0 to 0x10ffff, and that it is a "one-line description". So maybe this is actually a bug in Erlang/OTP? As far as I can tell, it should accept "ã" as a description, given the specification.

**Comment by lexmag** [comment]
I strongly believe this is a problem with `:epp.read_encoding_from_binary/1` function.

**Comment by antipax** [comment]
Ah, now I see how one of those "bad encodings" gets returned. I think you're right.

**Comment by lexmag** [comment]
There is a code snippet to reproduce the issue:
```elixir
app_file = :code.where_is_file('my_app.app')
{:ok, bin, _} = :erl_prim_loader.get_file(app_file)
:unicode.characters_to_list(bin, :utf8)
```
returns:
```elixir
{:error,
 '{application,my_app,\n             [{registered,[]},\n              {description,"S',
 <<227, 111, 34, 125, 44, 10, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 123, 118, 115, 110, 44, 34, 48, 46, 48, 46, 49, 34, 125, 44, 10, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, ...>>}
```

**Comment by josevalim** [comment]
@lexmag we can see in your comment that the codepoint 227 is being kept as is instead of being represented with the bytes <<195, 163>>. We need to find why is that.

**Comment by josevalim** [comment]
Ok, I found out the root cause. Thanks everyone. I will push a fix soon. :)

**Comment by josevalim** [comment]
Here we go: https://github.com/elixir-lang/elixir/commit/2d4a4b1ad5d5f60128e32121fba01388af4dfb7b :)

**Pull request closed without merging**